project/CrossJava.scala

```diff
@@ -0,0 +1,135 @@
+/**
+  * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
+  */
+
+package akka
+
+import java.io.File
+
+import sbt._
+
+import scala.collection.immutable.ListMap
+
+case class JavaVersion(numbers: Vector[Long], vendor: Option[String]) {
+  def numberStr: String = numbers.mkString(".")
+
+  def withVendor(vendor: Option[String]) = copy(vendor = vendor)
+  def withVendor(vendor: String) = copy(vendor = Option(vendor))
+  def withNumbers(numbers: Vector[Long]) = copy(numbers = numbers)
+
+  override def toString: String = {
+    vendor.map(_ + "@").getOrElse("") + numberStr
+  }
+}
+object JavaVersion {
+  def apply(version: String): JavaVersion = CrossJava.parseJavaVersion(version)
+  def apply(numbers: Vector[Long], vendor: String): JavaVersion = new JavaVersion(numbers, Option(vendor))
+}
+
+
+/** Tools for discovering different Java versions,
+  * will be in sbt 1.1.7 (https://github.com/sbt/sbt/pull/4139) but until that time replicated here */
+object CrossJava {
+  object Keys {
+    val discoveredJavaHomes = settingKey[Map[String, File]]("Discovered Java home directories")
```

Comment:
tried this on my Mac:
```
ListMap(9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 1.7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home, 1.8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home)
```

However, I also have a newer `jdk1.8.0_121`.
ls /Library/Java/JavaVirtualMachines
```
jdk-9.0.4.jdk		jdk1.7.0_45.jdk		jdk1.8.0.jdk		jdk1.8.0_121.jdk	jdk1.8.0_45.jdk
```

Reply:
Ow, for me on Debian it is:
```
ListMap(9 -> /usr/lib/jvm/java-9-openjdk-amd64, 8 -> /usr/lib/jvm/java-8-openjdk-amd64, 10 -> /usr/lib/jvm/java-10-openjdk-amd64)
```
so notably for me the key is '8', not '1.8'.

As noted in the comment I temporarily took this code from SBT. @2m do you know what the desired format was? I guess that should be fixed both here and in SBT?

Reply:
It should actually accept both: https://github.com/sbt/sbt/commit/aff9e0110cd7f4b942e194e1eb979f89fbf64f1f#diff-d0af389abdaf608fbd5bbd98420b2e27

Reply:
Ah, but that only goes for `fullJavaHomes`, not `discoveredJavaHomes` - I should've used `fullJavaHomes` in any case I suppose.

Reply:
Yup. `fullJavaHomes` is the discovered plus user overrides. Similar to `resolvers` and `fullResolvers`.

Reply:
It's still not finding my newer `jdk1.8.0_121`

```
sbt:akka-actor> show discoveredJavaHomes
[info] ListMap(9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 1.7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home, 1.8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home)

sbt:akka-actor> show fullJavaHomes
[info] Map(1.9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home, 9 -> /Library/Java/JavaVirtualMachines/jdk-9.0.4.jdk/Contents/Home, 1.7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home, 1.8 -> /Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk/Contents/Home, 7 -> /Library/Java/JavaVirtualMachines/jdk1.7.0_45.jdk/Contents/Home)
```

```
ls /Library/Java/JavaVirtualMachines
jdk-9.0.4.jdk		jdk1.7.0_45.jdk		jdk1.8.0.jdk		jdk1.8.0_121.jdk	jdk1.8.0_45.jdk
```

Reply:
Ah yes, I'll look into that one. Looks like it just picks a random one right now.

Reply:
@patriknw should we prefer jdk1.8.0_121.jdk to jdk1.8.0.jdk Not sure if the latter is a symlink or more like `jdk1.8.0_0.jdk`?

Reply:
no symlink, `jdk1.8.0.jdk` is a really old one. I'm fine with removing that one and you can assume that they have a `_version` that can be sorted